### PR TITLE
Remove Mypy type errors

### LINKED ISSUE #688: Run mypy in run_all_tests.sh
Ideally we want both type checkers to work correctly at least for the library. Also mypy should offer a speedup

_Originally posted by @jheek in https://github.com/google/flax/discussions/685#discussioncomment-140141_

### PR DESCRIPTION
**Pull request by zaxtax**
# What does this PR do?

This PR makes the annotation changes necessary so that mypy checks would all pass under CI

Fixes #688

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other
      checks if that's the case).
- [x] This change is discussed in a Github issue/
      [discussion](https://github.com/google/flax/discussions) (please add a
      link).
- [ ] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [ ] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: flax/core/scope.py [hunk 1]
```diff
@@ -619,8 +619,9 @@ def is_collection_empty(self, col: str) -> bool:
   def _mutable_collection(self, col: str) -> MutableCollection:
     """Returns the collection `col` as a mutable object."""
     assert self.is_mutable_collection(col), f'Collection {col} is not mutable'
+
     if col not in self._variables:
-      if self.parent:
+      if self.parent and self.name is not None:
```
Comment (cgarciae) [inline flax/core/scope.py:624]: I think its safer to use an assert over `name` so the runtime behaviour remains the same:
```python
if self.parent:
  assert self.name is not None
  ...
```

File: flax/core/scope.py [hunk 2]
```diff
@@ -837,7 +838,7 @@ def bind(variables: VariableDict,
 
 def apply(fn: Callable[..., Any],
           mutable: CollectionFilter = False,
-          flags: Optional[Dict] = None) -> Callable[..., Any]:
+          flags: Optional[FrozenDict] = None) -> Callable[..., Any]:
```
Comment (cgarciae) [inline flax/core/scope.py:841]: We seem to accept both `FrozenDict` and `Dict`, probably the correct type is either `Union[Dict, FrozenDict]` or even `Mapping`.

Comment (zaxtax) [reply flax/core/scope.py:841]: I think changing to `Mapping` is best

File: flax/core/scope.py [hunk 3]
```diff
@@ -632,7 +634,7 @@ def _mutable_collection(self, col: str) -> MutableCollection:
   def _collection(self, col: str) -> Collection:
     """Returns a collection of variables of collection `col`."""
     if col not in self._variables:
-      if self.parent:
+      if self.parent and self.name is not None:
```
Comment (cgarciae) [inline flax/core/scope.py:637]: Same issue as before, lets use an assert here.

**Comment by levskaya** [comment]
I'm seeing a ton of failures internally that all seem like duplicates of:
```
Base signature: 'def builtins.object.__dir__(self) -> List[str]'.
  Subclass signature: 'def flax.linen.module.Module.__dir__(self) -> Iterable[str]'.
  Return type mismatch.
```

**Comment by cgarciae** [comment]
Thanks @levskaya! Makes sense, we'll revert that.

**Comment by zaxtax** [comment]
> I'm seeing a ton of failures internally that all seem like duplicates of:
> 
> ```
> Base signature: 'def builtins.object.__dir__(self) -> List[str]'.
>   Subclass signature: 'def flax.linen.module.Module.__dir__(self) -> Iterable[str]'.
>   Return type mismatch.
> ```

Thanks for the heads-up. If it's at all possible, @levskaya is there a minimal example that triggers the error you can share? @cgarciae does that seem like a reasonable request? Since, I think if that's the only issue, we might be able to not revert too much and still get something modest merged in.

**Comment by cgarciae** [comment]
@zaxtax oh, I meant just revert the signature for `Module.__dir__`'s output from `Iterable` to `List`.

File: flax/core/lift.py [hunk 4]
```diff
@@ -1264,8 +1264,8 @@ def jit(fn: Callable[..., Any],
   # Close over scope_fn & repack_fn to avoid recompilation
   # this is impure but we use the fingerprint arg to differentiate between cases
   # where scope_fn or repack_fn actually produce non-identical results.
-  scope_fn = None  # type: Callable
-  repack_fn = None  # type: Callable
+  scope_fn = None  # type: Optional[Callable]
```
Comment (levskaya) [inline flax/core/lift.py:1267]: Curious: why aren't we using inline actual type annotations here?

**Pull request merged**